### PR TITLE
[SPARK-30556][SQL] Copy sparkContext.localproperties to child thread inSubqueryExec.executionContext

### PR DESCRIPTION
**Pull request by ajithme**
### What changes were proposed in this pull request?
In `org.apache.spark.sql.execution.SubqueryExec#relationFuture` make a copy of `org.apache.spark.SparkContext#localProperties` and pass it to the sub-execution thread in `org.apache.spark.sql.execution.SubqueryExec#executionContext`

### Why are the changes needed?
Local properties set via sparkContext are not available as TaskContext properties when executing  jobs and threadpools have idle threads which are reused

Explanation:
When `SubqueryExec`, the relationFuture is evaluated via a separate thread. The threads inherit the `localProperties` from `sparkContext` as they are the child threads.
These threads are created in the `executionContext` (thread pools). Each Thread pool has a default keepAliveSeconds of 60 seconds for idle threads.
Scenarios where the thread pool has threads which are idle and reused for a subsequent new query, the thread local properties will not be inherited from spark context (thread properties are inherited only on thread creation) hence end up having old or no properties set. This will cause taskset properties to be missing when properties are transferred by child thread via `sparkContext.runJob/submitJob`

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by ajithme** [comment]
cc @dongjoon-hyun @srowen @HyukjinKwon @gatorsmile @cloud-fan

**Comment by dongjoon-hyun** [comment]
ok to test

**Review by dongjoon-hyun — COMMENTED** [review]
For WIP PR, please add `[WIP]` into the PR title.
> ### How was this patch tested?
> WIP

**Comment by srowen** [comment]
Closely related to https://github.com/apache/spark/pull/27266?

**Comment by ajithme** [comment]
> Closely related to #27266?

Currently, `org.apache.spark.sql.execution.SubqueryExec#executionContext` has a hardcoded size of 16 threads which makes writing UT difficult so i plan to make it configurable + fix subquery thread bug hence created a new pull request compared to broadcast bug (which is independent of this fix) in https://github.com/apache/spark/pull/27266

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -749,9 +749,11 @@ case class SubqueryExec(name: String, child: SparkPlan)
   private lazy val relationFuture: Future[Array[InternalRow]] = {
     // relationFuture is used in "doExecute". Therefore we can get the execution id correctly here.
     val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    val localProps = Utils.cloneProperties(sparkContext.getLocalProperties)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Shouldn't we fix this in `SQLExecution.withExecutionId`? That way we can eliminate the issue you are also trying to fix in https://github.com/apache/spark/pull/27267.

**Comment by hvanhovell** [comment]
IMO the thread pools should not be a big issue. The subquery is guaranteed to be executed on a different thread (you can even add an assert for this). You just set some unique property on the local properties (value should also be unique), construct the something that contains a broadcast join, use an accumulator that you modify using either a UDF (easy) or a dataset operation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: When it execute the Future object in `SQLExecution.withExecutionId` the thread from the pool has already lost the local properties so it wont work

**Comment by ajithme** [comment]
> IMO the thread pools should not be a big issue. The subquery is guaranteed to be executed on a different thread (you can even add an assert for this). You just set some unique property on the local properties (value should also be unique), construct the something that contains a broadcast join, use an accumulator that you modify using either a UDF (easy) or a dataset operation.

Agree. But with a pool size of 16, i will have to ensure all 16 threads are used at-least once and are alive to reproduce this issue and not to make that test flaky. Its lot easier if i can set the pool size to 1 in test to reproduce.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 2]
```diff
@@ -749,9 +749,11 @@ case class SubqueryExec(name: String, child: SparkPlan)
...
     Future {
       // This will run in another thread. Set the execution id so that we can connect these jobs
       // with the correct execution.
+      sparkContext.setLocalProperties(localProps)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: what's the consequence if we don't do this? Do you run a UDF in subquery which accesses task context properties?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Fair point. I prefer creating a narrow waist here. How about you add a method to SQLExecution that creates the future and sets up the local properties?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 2 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: yes, if the subquery output depends on the task properties, it will affect i have added a UT case to simulate the behaviour

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Yes, seems good approach. let me try to modify as per your suggestion

**Comment by hvanhovell** [comment]
@ajithme why do we need to make sure all 16 threads are used?

**Comment by cloud-fan** [comment]
I think this is a common problem when we run code in a thread pool. I just realized that we hit a similar issue before with `SQLConf.get` and we add a hack to work around it.

Shall we add a util function to easily capture the necessary thread locals and propagate them to thread pool? For example, we can add a `SQLExecution.withThreadLocalCaptured`
```
def withThreadLocalCaptured(f: => T): Future[T] = {
  val activeSession = ...
  val localProperties = ...
  // any other important thread locals?
  Future {
    // set active sessionn
    // set local properties
   f
  }
}
```

**Comment by ajithme** [comment]
> util function to easily capture the necessary thread locals and propagate them to thread pool? For example, we can add a `SQLExecution.withThreadLocalCaptured`

sure @cloud-fan , i think similar approach was proposed by @hvanhovell , i will update my PR to introduce a utility than having a local fix.

**Comment by ajithme** [comment]
cc @cloud-fan @hvanhovell  i have updated PR as per the suggestion. Please review

**Comment by ajithme** [comment]
This seems to be problems what i try to fix in
1. `BroadcastExchangeExec` (Refer: https://github.com/apache/spark/pull/27266)
2. `SubqueryBroadcastExec` 
3.  `SubqueryExec` (Refer: https://github.com/apache/spark/pull/27267)
so should i try to fix all in one PR or should i have a separate PR for each.? i previously raised them separately so that i can complete them with individual UTs.

Please suggest @cloud-fan @srowen @dongjoon-hyun @hvanhovell, i am kinda neutral if to make separate PRs or single PR

**Comment by cloud-fan** [comment]
I think we can fix one place in this PR, and send more PRs to fix more places later, with tests.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 3]
```diff
@@ -164,4 +166,18 @@ object SQLExecution {
       }
     }
   }
+
+  /**
+   * Wrap passed function to ensure sparkContext local properties are forwarded to execution thread
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: `... ensure necessary thread-local variables like SparkContext local properties are forwarded ...`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 4]
```diff
@@ -164,4 +166,18 @@ object SQLExecution {
...
+   */
+  def withThreadLocalCaptured[T](sparkSession: SparkSession)(body: => T)(
+    exec: ExecutionContext): Future[T] = {
+    val activeSession = sparkSession
+    val localProps = Utils.cloneProperties(sparkSession.sparkContext.getLocalProperties)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: do we have to clone? I know the local properties are mutable, but that was the behavior until now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 3 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Updated

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 4 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: This is just to ensure case when the future is in waiting due to `ExecutorService` threads being fully occupied and meanwhile any changes happen in localProperties, that do not affect the properties which were passed on submission. Do you feel its reasonable.?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Yea I think it's reasonable, but a little worried about performance. @hvanhovell what do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 5]
```diff
@@ -164,4 +166,19 @@ object SQLExecution {
       }
     }
   }
+
+  /**
+   * Wrap passed function to ensure necessary thread-local variables like
+   * SparkContext local properties are forwarded to execution thread
+   */
+  def withThreadLocalCaptured[T](sparkSession: SparkSession)(body: => T)(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: shall we define it as `(sparkSession: SparkSession, exec: ExecutionContext)(body: => T)`?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 6]
```diff
@@ -125,6 +125,39 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+        .createOrReplaceTempView("l")
+      Seq(true)
+        .toDF()
+        .mapPartitions { _ =>
+          val conf = SQLConf.get
+          conf.isInstanceOf[ReadOnlySQLConf] && conf.getConfString("spark.sql.y") == "e" match {
+            case true => Iterator(true)
+            case false => Iterator.empty
+          }
+        }
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: shall we create a method to define this df to save duplicated code?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 5 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: updated

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 6 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: i have updated the PR with generic method to create df, is it ok now..?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 4 above)
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: I would be in favor of cloning them (BTW I am not sure why Utils.cloneProperties does not just use clone - the current implementation is probably less efficient). My main worry is that the local properties object is shared between multiple threads (which can happen if have multiple broadcast joins for example), and that these threads no start influencing each other and the main thread; I would like to avoid midnight debugging sessions caused by this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 7]
```diff
@@ -164,4 +166,19 @@ object SQLExecution {
...
+  /**
+   * Wrap passed function to ensure necessary thread-local variables like
+   * SparkContext local properties are forwarded to execution thread
+   */
+  def withThreadLocalCaptured[T](sparkSession: SparkSession, exec: ExecutionContext)(
+    body: => T): Future[T] = {
+    val activeSession = sparkSession
+    val localProps = Utils.cloneProperties(sparkSession.sparkContext.getLocalProperties)
+    Future {
+      SparkSession.setActiveSession(activeSession)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Can you please unset/restore both of them to their old state?

Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Do you mean i should clear the thread locals.? as these are worker threads, they do not have any old state as such. For example, when the thread is created first time, it will inherit from parent thread, but when resued, restoring would cause the previous properties to be set which is stale. IMO, setting this on every execution should suffice. Please correct me if wrong

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 8]
```diff
@@ -125,6 +125,41 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
     val e = intercept[SparkException](dummyQueryExecution1.toRdd.collect())
     assert(e.getCause.isInstanceOf[NoSuchElementException])
   }
+
+  test("SPARK-30556 propagate local properties to subquery execution thread") {
+    withSQLConf("spark.sql.subquery.maxThreadThreshold" -> "1") {
+      Seq(true)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: we should wrap the test with `withTempView`, which removes the temp view at the end.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 9]
```diff
@@ -148,6 +148,14 @@ object StaticSQLConf {
       .checkValue(thres => thres > 0 && thres <= 128, "The threshold must be in [0,128].")
       .createWithDefault(128)
 
+  val SUBQUERY_MAX_THREAD_THRESHOLD =
+    buildStaticConf("spark.sql.subquery.maxThreadThreshold")
+      .internal()
+      .doc("The maximum degree of parallelism to execute the subquery. ")
+      .intConf
+      .checkValue(thres => thres > 0 && thres <= 128, "The threshold must be in [0,128].")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: `[0,128]` -> `(0, 128]` because 0 is not in the range.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: You may want to change line 148, too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 10]
```diff
@@ -164,4 +166,19 @@ object SQLExecution {
...
     }
   }
+
+  /**
+   * Wrap passed function to ensure necessary thread-local variables like
+   * SparkContext local properties are forwarded to execution thread
+   */
+  def withThreadLocalCaptured[T](sparkSession: SparkSession, exec: ExecutionContext)(
+    body: => T): Future[T] = {
+    val activeSession = sparkSession
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Shall we have `val sc = sparkSession.sparkContext` here?

Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 9 above)
Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 8 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: updated

**Comment by ajithme** [comment]
Updated as per comments @cloud-fan @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 11]
```diff
@@ -125,6 +125,44 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
     val e = intercept[SparkException](dummyQueryExecution1.toRdd.collect())
     assert(e.getCause.isInstanceOf[NoSuchElementException])
   }
+
+  test("SPARK-30556 propagate local properties to subquery execution thread") {
+    withSQLConf("spark.sql.subquery.maxThreadThreshold" -> "1") {
+      withTempView("l", "m", "n") {
+        Seq(true)
+          .toDF()
+          .createOrReplaceTempView("l")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: super nit: can we put the above three lines into one line?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 12]
```diff
@@ -125,6 +125,44 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+        val confKey = "spark.sql.y"
+
+        def createDataframe(confKey: String, confValue: String): Dataset[Boolean] = {
+          Seq(true)
+            .toDF()
+            .mapPartitions { _ =>
+              val conf = SQLConf.get
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: can we test with `TastContext.get.getLocalProperty` directly?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 13]
```diff
@@ -125,6 +125,44 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+                case true => Iterator(true)
+                case false => Iterator.empty
+              }
+            }
+        }
+
+        // set local configuration and assert
+        val confValue1 = "e"
+        createDataframe(confKey, confValue1)
+          .createOrReplaceTempView("m")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: super nit: can we put the above two lines into one line?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 14]
```diff
@@ -125,6 +125,44 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+        spark.sparkContext.setLocalProperty(confKey, confValue1)
+        assert(sql("select * from l where exists (select * from m)").collect.size == 1)
+
+        // change the conf value and assert again
+        val confValue2 = "f"
+        createDataframe(confKey, confValue2)
+          .createOrReplaceTempView("n")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 15]
```diff
@@ -164,4 +166,20 @@ object SQLExecution {
       }
     }
   }
+
+  /**
+   * Wrap passed function to ensure necessary thread-local variables like
+   * SparkContext local properties are forwarded to execution thread
+   */
+  def withThreadLocalCaptured[T](sparkSession: SparkSession, exec: ExecutionContext)(
+    body: => T): Future[T] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: the indentation looks weird. How about
```
def withThreadLocalCaptured[T](
    sparkSession: SparkSession, exec: ExecutionContext)(body: => T): Future[T]
```

**Review by cloud-fan — APPROVED** [review]
LGTM except a few code style comments. Thanks for the fix!

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 11 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 13 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: Yes, i have updated

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 12 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 14 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: made single line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 15 above)
Comment (ajithme) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: Ok, weird style was because of `scalafmt`, now as per suggestion i have manually formatted and updated the PR

**Comment by ajithme** [comment]
@cloud-fan Updated the style as per your suggestion

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 16]
```diff
@@ -145,9 +145,17 @@ object StaticSQLConf {
...
         "cause memory problem.")
       .intConf
-      .checkValue(thres => thres > 0 && thres <= 128, "The threshold must be in [0,128].")
+      .checkValue(thres => thres > 0 && thres <= 128, "The threshold must be in (0,128].")
       .createWithDefault(128)
 
+  val SUBQUERY_MAX_THREAD_THRESHOLD =
+    buildStaticConf("spark.sql.subquery.maxThreadThreshold")
+      .internal()
+      .doc("The maximum degree of parallelism to execute the subquery. ")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: Remove the redundant space; `. "` -> `."`?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 17]
```diff
@@ -125,6 +125,38 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
     val e = intercept[SparkException](dummyQueryExecution1.toRdd.collect())
     assert(e.getCause.isInstanceOf[NoSuchElementException])
   }
+
+  test("SPARK-30556 propagate local properties to subquery execution thread") {
+    withSQLConf("spark.sql.subquery.maxThreadThreshold" -> "1") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: Shall we use `SUBQUERY_MAX_THREAD_THRESHOLD` instead of `"spark.sql.subquery.maxThreadThreshold"`? That's better when we change the configuration string name later.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 18]
```diff
@@ -125,6 +125,38 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+                case false => Iterator.empty
+              }
+            }
+        }
+
+        // set local configuration and assert
+        val confValue1 = "e"
+        createDataframe(confKey, confValue1).createOrReplaceTempView("m")
+        spark.sparkContext.setLocalProperty(confKey, confValue1)
+        assert(sql("select * from l where exists (select * from m)").collect.size == 1)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: If you don't mind, let's capitalize the SQL keywords like `SELECT * FROM l WHERE EXISTS (SELECT * FROM m)`.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 19]
```diff
@@ -125,6 +125,38 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+
+        // change the conf value and assert again
+        val confValue2 = "f"
+        createDataframe(confKey, confValue2).createOrReplaceTempView("n")
+        spark.sparkContext.setLocalProperty(confKey, confValue2)
+        assert(sql("select value from l where exists (select * from n)").collect().size == 1)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: ditto.

Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 18 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 17 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 16 above)
Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: typo. Corrected

**Comment by ajithme** [comment]
@dongjoon-hyun Updated with fix

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.

**Comment by dongjoon-hyun** [comment]
Could you make a backporting PR against branch-2.4, @ajithme ?

**Comment by ajithme** [comment]
Thanks @dongjoon-hyun @cloud-fan @hvanhovell 

@dongjoon-hyun sure, i will make a backport PR for this over 2.4

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 7 above)
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: I think we should restore them to their previous state. In this particular case you can hold on to a SparkSession that is no longer in use.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala [hunk 20]
```diff
@@ -145,9 +145,17 @@ object StaticSQLConf {
         "cause longer waiting for other broadcasting. Also, increasing parallelism may " +
         "cause memory problem.")
       .intConf
-      .checkValue(thres => thres > 0 && thres <= 128, "The threshold must be in [0,128].")
+      .checkValue(thres => thres > 0 && thres <= 128, "The threshold must be in (0,128].")
       .createWithDefault(128)
 
+  val SUBQUERY_MAX_THREAD_THRESHOLD =
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: This is actually a different change right?

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: It should also be a static configuration since we can only change it at startup.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala [hunk 21]
```diff
@@ -125,6 +125,38 @@ class ExecutorSideSQLConfSuite extends SparkFunSuite with SQLTestUtils {
...
+            .mapPartitions { _ =>
+              TaskContext.get.getLocalProperty(confKey) == confValue match {
+                case true => Iterator(true)
+                case false => Iterator.empty
+              }
+            }
+        }
+
+        // set local configuration and assert
+        val confValue1 = "e"
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: IMO it is better to use something unique here to avoid a fluke. How about using `UUID.randomUUID().toString()`?

**Comment by hvanhovell** [comment]
@ajithme can you address my comments in a follow up?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 20 above)
Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: 1. This is just opening as a configration to make the change testable

2. This is part of `StaticSQLConf` which is defined at startup, is there any other mechanism to define static conf.?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala (hunk 21 above)
Comment (ajithme) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/ExecutorSideSQLConfSuite.scala]: I am neutral about both approaches. Just wanted to have a fixed input value to get predictable output value from testcase. I can update this and raise followup if you insist

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 20 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: If its a static conf, then why isn't your unite test failing? Moreover, if its static then setting it in your test probably does not have any effect because we use the same JVM/SparkContext to run most tests, the chances are pretty high that it has been set before.

Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: i see. as per documentation in https://github.com/apache/spark/blob/705fc6ad9328c1092299b83071b6ec3b1d6f9c4d/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala#L26 its should not modified. I followed the same way `BroadcastExchangeExec` creates executionContext. My initial guess is executionContext is not created till first subquery hence it work for UT. i will further investigate and get back with analysis.

**Comment by gatorsmile** [comment]
@ajithme Can you submit a PR to address the comments?

**Comment by gatorsmile** [comment]
@xuanyuanking Could you submit a PR to address the comments?

**Comment by gatorsmile** [comment]
https://github.com/apache/spark/pull/27267/files#r370089158 is the major comment we need to address.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 20 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: I think we should submit a separate PR for this change.

**Comment by ajithme** [comment]
> @ajithme Can you submit a PR to address the comments?

@gatorsmile @xuanyuanking Sure, i will submit a follow up PR shortly.

**Comment by xuanyuanking** [comment]
@ajithme Sorry, I nearly submit a follow-up PR, could you help on reviewing? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala (hunk 20 above)
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/StaticSQLConf.scala]: I think we could keep this in StaticSQLConf.

- I locally test the static config takes effect, it can only be set while startup.
- The UT can pass because it was used in lazy val `SubqueryExec.relationFuture` on the executor side, so the `withSQLConf` in UT could set the config before executor start.

cc @hvanhovell

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 22]
```diff
@@ -164,4 +166,20 @@ object SQLExecution {
...
+   * SparkContext local properties are forwarded to execution thread
+   */
+  def withThreadLocalCaptured[T](
+      sparkSession: SparkSession, exec: ExecutionContext)(body: => T): Future[T] = {
+    val activeSession = sparkSession
+    val sc = sparkSession.sparkContext
+    val localProps = Utils.cloneProperties(sc.getLocalProperties)
+    Future {
+      SparkSession.setActiveSession(activeSession)
+      sc.setLocalProperties(localProps)
```
Comment (dbaliafroozeh) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: @hvanhovell two questions:
- Shouldn't we clone`localProps` here? in the sense that what if a concurrent thread modify them?
- Does the order of setting `localProps` and `activeSession` matter?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: the `localProps` is already a clone: https://github.com/apache/spark/pull/27267/files#diff-ab49028253e599e6e74cc4f4dcb2e3a8R178

And I think the order doesn't matter.
